src/apprt/gtk/TabView.zig

```diff
@@ -90,10 +107,83 @@ pub fn asWidget(self: *TabView) *gtk.Widget {
     return self.tab_view.as(gtk.Widget);
 }
 
+pub fn syncAppearance(self: *TabView) void {
+    self.setPageIcons();
+}
+
 pub fn nPages(self: *TabView) c_int {
     return self.tab_view.getNPages();
 }
 
+fn setPageIcons(self: *TabView) void {
+    const count: usize = @intCast(self.nPages());
+    for (0..count) |position| {
+        const page = self.tab_view.getNthPage(@intCast(position));
+        const icon: ?*gio.Icon = icon: {
+            const number = position + 1;
+            if (!self.window.config.window_tab_icons) break :icon null;
+
+            if (gtk_version.atLeast(4, 12, 0)) {
+                const color = switch (self.window.app.core_app.config_conditional_state.theme) {
+                    .light => "black",
+                    .dark => "white",
+                };
+                var buf: [1024]u8 = undefined;
+                const svg = std.fmt.bufPrintZ(
+                    &buf,
+                    \\<svg width="48px" height="48px">
+                    \\  <circle
+                    \\    cx="50%" cy="50%"
+                    \\    r="23px"
+                    \\    stroke="{[color]s}"
+                    \\    stroke-width="2px"
+                    \\    fill="none"/>
+                    \\  <text
+                    \\    x="50%" y="50%"
+                    \\    dy=".35em"
+                    \\    text-anchor="middle"
+                    \\    dominant-baseline="central"
+                    \\    fill="{[color]s}"
+                    \\    font-size="{[size]d}px"
+                    \\    font-family="mono">{[number]d}</text>
+                    \\</svg>
+                    \\
+                ,
+                    .{
+                        .color = color,
+                        .size = @as(usize, switch (number) {
+                            0...9 => 36,
+                            10...99 => 30,
+                            else => 22,
+                        }),
+                        .number = number,
+                    },
+                ) catch unreachable;
+
+                const bytes = glib.Bytes.new(svg.ptr, svg.len);
+                defer bytes.unref();
+
+                break :icon gio.BytesIcon.new(bytes).as(gio.Icon);
+            }
```

Comment:
I think for clarity it's better if this is separated out into its own function. Bonus points if the SVG template can be used by the gresource generation code to generate the 10 default icons, so that we don't have to commit them to the repo :D

Reply:
After dropping the static SVGs the code is much cleaner. I don't think that we need to break this out into a separate function at this point unless we need to reuse the code somewhere else.